_source/_posts/2019-09-19-nodejs-typescript.md

```diff
@@ -0,0 +1,637 @@
+description: "Learn to build a simple blog server using Node.js and TypeScript"
+tags: [nodejs, typescript, nestjs, secure, oauth, oidc, javascript]
+tweets:
+- "Learn to build a blog application using Node.js and TypeScript with NestJS! →"
+- "Yes, you can use TypeScript with Node.js! Here's an example of building a blog application using Node.js and TypeScript with NestJS! →"
+- "Can you use TypeScript with Node.js? Absolutely! Here's one approach to building a blog application with TypeScript and NestJS! →"
+image: blog/featured/okta-node-bottle-headphones.jpg
+---
+
+As dynamically typed languages became prominent during the last decade, typeless (or should I say lawless?) programming became the norm for the backend as well as the frontend.
+
+Many people believe the simplicity of "just writing" code is efficient for providing a proof of concept or prototyping applications.
+
+However, as those applications grow, the typeless code used to build them often becomes incredibly convoluted and more difficult (some would say impossible) to manage.
+
+In the worst cases, developers end up resorting to testing and prayer in the hopes that simple operations like renaming a function won't break the whole complex system.  However, developers can build their code from day one to make their codebase readable at scale.
+
+## TypeScript to the rescue
+
+The key to making a large (or even small!) codebase readable is to keep track of all the data types flowing through the app.
+
+To help developers with this, TypeScript provided a typed superset of plain old JavaScript.
+This means that a regular JavaScript developer's existing knowledge applies to TypeScript.
+
+There are other efficient ways developers can use their knowledge and code in new ways. Developers who use NodeJS can now recycle their frontend code for the backend, and vice versa.
+
+Additionally, the introduction of static typing to JavaScript allows developers to reuse some of that old Java/C# knowledge to help maintain their NodeJS codebases.
+
+## Today's Project: A Simple Blogging Engine
+
+In this tutorial, we need a new backend for a blogging engine. This app will contain a list of blog posts, each with a title, content, and author.
+
+The authors will be able to register and log in using the session-in-cookies approach. Okta will manage the users, and all blog posts will be saved in memory.
+
+Documentation for the backend will be exposed as a Swagger specification generated from TypeScript typings for classes and methods.
+
+## Dependencies for the NodeJS Typescript Sample App
+
+- NestJS to bootstrap the server
+- Swagger to define the server docs
+- Okta for user management and authentication
+- NodeJS
+- `validate-class` npm library for input validation
+- nodemon to auto-restart the server on every change
+
+## Initialize
+
+Create an empty directory `bloggeur`, `cd` into it, and type `npm init` there. After pressing the `ENTER` key  a few times, a NodeJS project will initiate.
```

Comment:
Directory name looks French. I don't have a problem with it, just seems strange.